certora/specs/Account.spec

```
@@ -0,0 +1,488 @@
+    {
+        preserved uninstallModule(uint256 moduleTypeId, address otherModule, bytes deInitData) with (env e) {
+            requireInvariant consistencyIndexExecutor(index);
+            requireInvariant consistencyKeyExecutor(module);
+            requireInvariant atUniquenessExecutor(index, require_uint256(_executorLength() - 1));
+            requireInvariant cleanStorageExecutor(require_uint256(_executorLength() - 1));
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│                                                  Module management                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+// This guarantees that at most one fallback module is active for a given initData (i.e. selector)
+rule fallbackModule(address module, bytes initData) {
+    assert isModuleInstalled(3, module, initData) <=> getFallbackHandler(getDataSelector(initData)) == module;
+}
+
+rule moduleManagementRule(env e, method f, calldataarg args, uint256 moduleTypeId, address module, bytes additionalContext)
+    filtered { f -> !f.isView }
+{
+    bytes context;
+    require context.length == 0;
+
+    bool isEntryPoint = isEntryPoint(e);
+    bool isEntryPointOrSelf = isEntryPointOrSelf(e);
+    bool isExecutionModule = isExecutionModule(e, context);
+
+    bool isModuleInstalledBefore = isModuleInstalled(moduleTypeId, module, additionalContext);
+    f(e, args);
+    bool isModuleInstalledAfter = isModuleInstalled(moduleTypeId, module, additionalContext);
+
+    assert (
+        isModuleInstalledBefore != isModuleInstalledAfter
+    ) => (
+        (
+            f.selector == sig:execute(bytes32,bytes).selector  &&
+            isEntryPointOrSelf
+        ) || (
+            f.selector == sig:executeFromExecutor(bytes32,bytes).selector &&
+            isExecutionModule
+        ) || (
+            f.selector == sig:installModule(uint256,address,bytes).selector &&
+            isEntryPointOrSelf
+        ) || (
+            f.selector == sig:uninstallModule(uint256,address,bytes).selector &&
+            isEntryPointOrSelf
+        )
+    );
+}
+
+rule installModuleRule(env e, uint256 moduleTypeId, address module, bytes initData) {
+    uint256 otherModuleTypeId;
+    address otherModule;
+    bytes otherInitData;
+
+    require moduleLengthSanity();
+    requireInvariant consistencyKeyExecutor(module);
+    requireInvariant consistencyKeyValidator(module);
+    requireInvariant consistencyKeyExecutor(otherModule);
+    requireInvariant consistencyKeyValidator(otherModule);
+
+    bool isModuleInstalledBefore = isModuleInstalled(moduleTypeId, module, initData);
+    bool isOtherModuleInstalledBefore = isModuleInstalled(otherModuleTypeId, otherModule, otherInitData);
+
+    installModule(e, moduleTypeId, module, initData);
```

Same comment as `moduleManagementRule`

```suggestion
    installModule@withrevert(e, moduleTypeId, module, initData);
```

I think the rule will become more complex since we may need to summarize the `isModuleType` function in the modules

---

same answer